orbmain/src/main/java/com/sun/corba/ee/impl/io/IIOPInputStream.java

```diff
@@ -1152,6 +1153,32 @@ private synchronized Object inputObject(Class clz, ClassInfoCache.ClassInfo cinf
         return currentObject;
     }
 
+    // This is a horrible workaround for a problem that looks a bit challenging to fix now.
+    // In JDK9, the Date class was modified to call defaultWriteObject(), which doesn't actually do anything but
+    // set the "defaultWriteObjectCalled" flag in the serialization of the class; unfortunately, the
+    // deserialization logic cannot handle that case and throws an exception. We are therefore employing this
+    // workaround to allow Date objects to be read across the wire between JDK8 and later JDKs.
+    // Fixing it will involve massive code simplification in order to eliminate duplicate code and allow for
+    // implementation of a simpler state machine that depends on the data sent rather than how it was produced.
+    private boolean readDefaultWriteObjectCalledFlag() throws IOException {
+        boolean sentDefaultWriteObjectCalled = readBoolean();
+
+        if (isDateClassWorkaroundRequired()) return getDateClassWorkaroundValue();
+        return sentDefaultWriteObjectCalled;
+    }
+
+    private boolean isDateClassWorkaroundRequired() {
+        return currentClassDesc.getName().equals(Date.class.getName());
+    }
+
+    private boolean getDateClassWorkaroundValue() {
+        return isJdk9_orLater();
```

Comment:
any reason that this just passes through to another private function. Couldn't the implementation of isJDK9 just be added here?

Reply:
It is intentional, in order to make it clear what is happening. The implementation of isJDK9 isn't obvious, so I use the name of the method to shorten the semantic distance between the implementation and the purpose. I have updated the code, hoping to improve readability.

I expect the reader to conclude:
- that the workaround is to replace the actual value of the flag with the workaround value
- that the workaround value is based on whether the code is running with code earlier than JDK9, when the change was made to the Date class.